job/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -243,16 +247,32 @@ private void copyLogs() {
                 AnnotatedLargeText<? extends FlowNode> logText = la.getLogText();
                 try {
                     long old = entry.getValue();
-                    long revised = logText.writeRawLogTo(old, listener.getLogger());
-                    if (revised != old) {
-                        entry.setValue(revised);
-                        modified = true;
+                    OutputStream logger;
+
+                    String prefix = getLogPrefix(node);
+                    if (prefix != null) {
+                        logger = new LogLinePrefixOutputFilter(listener.getLogger(), "[" + prefix + "] ");
+                    } else {
+                        logger = listener.getLogger();
                     }
-                    if (logText.isComplete()) {
-                        logText.writeRawLogTo(entry.getValue(), listener.getLogger()); // defend against race condition?
-                        assert !node.isRunning() : "LargeText.complete yet " + node + " claims to still be running";
-                        it.remove();
-                        modified = true;
+
+                    try {
+                        long revised = logText.writeRawLogTo(old, logger);
+                        if (revised != old) {
+                            entry.setValue(revised);
+                            modified = true;
+                        }
+                        if (logText.isComplete()) {
+                            logText.writeRawLogTo(entry.getValue(), logger); // defend against race condition?
+                            assert !node.isRunning() : "LargeText.complete yet " + node + " claims to still be running";
+                            it.remove();
+                            modified = true;
+                        }
+                    } finally {
+                        if (prefix != null) {
+                            // It's a LogLinePrefixOutputFilter (see above). Force an EOL by closing.
+                            logger.close();
```

Comment:
Would be more self-explanatory to call `LineTransformationOutputStream.forceEol` directly. I know `close` delegates to this, but you then have to dig into core sources to understand that.